unit_testing/cli/test_cli_sct_resample.py

```diff
@@ -1,14 +1,26 @@
-from pytest_console_scripts import script_runner
+
 import pytest
 import logging
 
+from spinalcordtoolbox.image import Image
+from spinalcordtoolbox.scripts import sct_resample
+
 logger = logging.getLogger(__name__)
 
 
-@pytest.mark.script_launch_mode('subprocess')
-def test_sct_resample_backwards_compat(script_runner):
-    ret = script_runner.run('sct_testing', '--function', 'sct_resample')
-    logger.debug(f"{ret.stdout}")
-    logger.debug(f"{ret.stderr}")
-    assert ret.success
-    assert ret.stderr == ''
+@pytest.mark.sct_testing
+@pytest.mark.usefixtures("run_in_sct_testing_data_dir")
+@pytest.mark.parametrize("path_in,type_arg,dim,expected_dim", [
+    # FIXME: The first test fails (and was also failing in 'sct_testing') -- Is it correct to expect '2.2'?
+    ('dmri/dmri.nii.gz', '-f', '0.5x0.5x1', (20, 21, 5, 7, 1.6826923, 1.6826923, 17.5, 2.2)),
+    ('t2/t2.nii.gz', '-mm', '0.97x1.14x1.2', (62, 48, 43, 1, 0.96774191, 1.1458334, 1.2093023, 1)),
+    ('t2/t2.nii.gz', '-vox', '120x110x26', (120, 110, 26, 1, 0.5, 0.5, 2.0, 1)),
+])
+def test_sct_resample_output_has_expected_dimensions(path_in, type_arg, dim, expected_dim):
+    """Run the CLI script and verify output file exists."""
+    path_out = 'resampled.nii.gz'
+    sct_resample.main(argv=['-i', path_in, '-o', path_out, type_arg, dim, '-v', '1'])
+    actual_dim = Image(path_out).dim
+
+    for actual_val, expected_val in zip(actual_dim, expected_dim):
+        assert actual_val == pytest.approx(expected_val)
```

Comment:
The first set of parameters fails. (See [these CI logs](https://travis-ci.com/github/neuropoly/spinalcordtoolbox/jobs/488247310#L1435-L1460).) They were copied as-is from the `sct_testing` test.

I'm not sure if this is an issue with `sct_resample` or the test itself. I would have to look more deeply into the meaning of the `.dim` attribute of images.

Reply:
the first test (dmri): the native images are:
~~~
dim1		40
dim2		42
dim3		5
dim4		7
pixdim1		0.841346
pixdim2		0.841346
pixdim3		17.500000
pixdim4		2.200000
~~~
so a downsampling by factor 2 along X and Y should indeed give:
~~~
(20, 21, 5, 7, 1.6826923, 1.6826923, 17.5, 2.2)
~~~
Now, the "2.2" on the field "pixdim4" corresponds to the MRI acquisition parameter "repetition time (TR)", so that's not something that changes here for the spatial interpolation. So there must be something wrong in the way this test was designed, because this should not be wrong:
https://travis-ci.com/github/neuropoly/spinalcordtoolbox/jobs/488247310#L1456
~~~
E           assert 1.0 == 2.2 ± 2.2e-06
E             +1.0
E             -2.2 ± 2.2e-06
~~~

Reply:
Thank you for the clarification! This answers all of my questions. :slightly_smiling_face:

Reply:
> So there must be something wrong in the way this test was designed, because this should not be wrong:

On further inspection, I think this test discovered a bug in `resampling.py :: resample_nib`. Because the failing test is a 4D image, here is the relevant section:

https://github.com/neuropoly/spinalcordtoolbox/blob/6d70ad82fbd9ba7961fc1ef7fbb7ebca70ac6fc7/spinalcordtoolbox/resampling.py#L111-L124

Here's a comparison between the original affine matrix and the resampled one:

```python
# Input 4D image
>>> image.get_affine()
array([[-8.40976596e-01,  0.00000000e+00, -5.18624723e-01,  9.03539562e+00],
       [-1.86897092e-03,  8.38979125e-01,  1.31120324e+00, -1.72721119e+01],
       [-2.48637367e-02, -6.30663186e-02,  1.74431019e+01, -2.61166191e+01],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00, 1.00000000e+00]])
>>> image.header.get_zooms()
(0.84134614, 0.84134614, 17.5, 2.2)

# Output 4D image
>>> img_r.get_affine()
array([[-1.68195319e+00,  0.00000000e+00, -5.18624723e-01, 9.03539562e+00],
       [-3.73794185e-03,  1.67795825e+00,  1.31120324e+00, -1.72721119e+01],
       [-4.97274734e-02, -1.26132637e-01,  1.74431019e+01, -2.61166191e+01],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,  1.00000000e+00]])
>>> img_r.header.get_zooms()
(1.6826923, 1.6826923, 17.5, 1.0)
```

I notice that `2.2` is not actually in either affine. Then I noticed this line in nibabel's documentation, re: [the sform affine](https://nipy.org/nibabel/nifti_images.html#the-sform-affine): 

> The header does not store the fourth row because it is always [0, 0, 0, 1] (see Coordinate systems and affines).

So, because the affine doesn't store TR, when we try and create a 4D image using 

https://github.com/neuropoly/spinalcordtoolbox/blob/6d70ad82fbd9ba7961fc1ef7fbb7ebca70ac6fc7/spinalcordtoolbox/resampling.py#L124

We lose the TR parameter.

Reply:
very nice catch! few things here:

>> The header does not store the fourth row because it is always [0, 0, 0, 1] (see Coordinate systems and affines).

>So, because the affine doesn't store TR (I think), when we try and create a 4D image using

No, that concerns something else. NIFTI convention adds an empty row just to be able to have a 4x4 matrix, which is easier to manipulate for linear algebra (vs. having a 3x4 matrix).  So: the TR is independent from this 4th row in the qform. However, it should _not_ be removed by our function, so that should be fixed 😊

Reply:
Thank you for the context! I feel like I should take some time to read through the nibabel documentation -- there is quite a lot of valuable information there. :slightly_smiling_face: 

> So: the TR is independent from this 4th row in the qform. However, it should not be removed by our function, so that should be fixed 

Aha, I see now. There's a corresponding `header.set_zooms` function alongside `header.get_zooms`. So, this should be a relatively easy fix. 

That's quite a gotcha, though! Passing an affine sets 3/4 zooms, but you have to manually set the 4th. Tricky.

Reply:
Opened an issue here: https://github.com/neuropoly/spinalcordtoolbox/issues/3294